network/src/network.rs

```diff
@@ -396,58 +407,96 @@ impl NetworkState {
+        allow_dial_to_self: bool,
+    ) -> Result<(), Error> {
+        if !self.can_dial(peer_id, &addr, allow_dial_to_self) {
+            return Err(Error::Custom(format!(
+                "ignore dialing peer_id {:?}, addr {}",
+                peer_id, addr
+            )));
         }
-        false
+
+        let peer_id_hash = Multihash::from_bytes(peer_id.as_bytes().to_vec())
+            .map_err(|err| Error::Custom(format!("peer_id is not valid: {}", err)))?;
+        addr.push(multiaddr::Protocol::P2p(peer_id_hash));
+        debug!("dialing {} with {:?}", addr, target);
+        p2p_control.dial(addr.clone(), target)?;
+        self.dialing_addrs
+            .write()
+            .insert(peer_id.to_owned(), Instant::now());
+        Ok(())
     }
 
     /// Dial just identify protocol
     pub fn dial_identify(&self, p2p_control: &ServiceControl, peer_id: &PeerId, addr: Multiaddr) {
-        self.dial(
+        if let Err(err) = self.dial_inner(
             p2p_control,
             peer_id,
             addr,
             DialProtocol::Single(IDENTIFY_PROTOCOL_ID.into()),
-        );
+            false,
+        ) {
+            debug!("dial_identify error: {}", err);
+        }
     }
 
     /// Dial just feeler protocol
     pub fn dial_feeler(&self, p2p_control: &ServiceControl, peer_id: &PeerId, addr: Multiaddr) {
-        if self.dial(
+        if let Err(err) = self.dial_inner(
             p2p_control,
             peer_id,
             addr,
             DialProtocol::Single(FEELER_PROTOCOL_ID.into()),
+            false,
         ) {
+            debug!("dial_feeler error {}", err);
+        } else {
             self.with_peer_registry_mut(|reg| {
                 reg.add_feeler(peer_id.clone());
             });
         }
     }
+
+    /// this method is intent to check advise addr by dial to self
+    pub(crate) fn try_dial_advise_addrs(&self, p2p_control: &ServiceControl) {
+        let mut pending_advise_addrs = self.pending_advise_addrs.write();
+        for addr in pending_advise_addrs.drain() {
+            trace!("try dial advise addr: {:?}", addr);
+            if let Err(err) = self.dial_inner(
+                p2p_control,
+                self.local_peer_id(),
+                addr,
+                DialProtocol::Single(IDENTIFY_PROTOCOL_ID.into()),
+                true,
+            ) {
+                debug!("try_dial_advise_addrs error {}", err);
+            }
+        }
+    }
+
+    pub fn new_advise_addrs(&self, iter: impl Iterator<Item = Multiaddr>) {
+        let mut listened_addrs = self.listened_addrs.write();
+        let mut pending_advise_addrs = self.pending_advise_addrs.write();
+        for addr in iter {
+            if let Some(score) = listened_addrs.get_mut(&addr) {
+                *score = score.saturating_add(1);
+                trace!(
+                    "increase score for exists advise addr: {:?} {}",
+                    addr,
+                    score
+                );
```

Comment:
Should remove address already in `pending_advise_addrs`

Reply:
There are exists verify little probabilities that when node receiving an advised addr, at the same time the node is dialing the same addr. I' ll handle this conflict in `ConnectToSelf` phase.